### DISCUSSION TITLE
📔【数据库】数据库的存储引擎与适用场景 🆕

### DISCUSSION THREAD

**Discussion by imageslr** [imageslr/blog#87] [Announcements]
https://imageslr.github.io/2020/db-engine.html 

 更多面试题总结请看：🗂【面试题】技术面试题汇总

**Comment by jingxuantju** [comment]
您好 我看到您文章中写的 “MyISAM 的数据是**顺序存储**的。索引的 B+ 树叶节点存放数据记录的地址，可以直接定位到数据，因此查找速度很快。”
这里的顺序存储是指按照主键的顺序么，如果是按照主键的顺序，那么按二级索引查是不是就不是顺序IO而是随机IO了呀？

**Comment by imageslr** [comment]
@jingxuantju 
1. MyISAM 是按照主键的顺序存储的。
2. MyISAM 不管是主键索引、还是二级索引，其叶节点存储的都是每个行记录的地址。拿这个地址，直接去磁盘读取对应的值。相当于每次都是随机 IO 吧。
3. InnoDB 则相反，叶节点里的所有记录保存在一个页面里，行记录之间都是链式存储的，所以没法直接读到一个行记录。必须直接加载整个页面进来，找到对应的行记录。
4. InnoDB 的主键索引和 MyISAM 也不太一样，前者的主键索引是聚集索引 (主键的叶节点包含了完整的数据，既是索引也是物理存储)，后者的主键索引只存储地址。

我也是网上查的，欢迎指正 :)

**Comment by jingxuantju** [comment]
> @jingxuantju
> 
> 1. MyISAM 是按照主键的顺序存储的。
> 2. MyISAM 不管是主键索引、还是二级索引，其叶节点存储的都是每个行记录的地址。拿这个地址，直接去磁盘读取对应的值。相当于每次都是随机 IO 吧。
> 3. InnoDB 则相反，叶节点里的所有记录保存在一个页面里，行记录之间都是链式存储的，所以没法直接读到一个行记录。必须直接加载整个页面进来，找到对应的行记录。
> 4. InnoDB 的主键索引和 MyISAM 也不太一样，前者的主键索引是聚集索引 (主键的叶节点包含了完整的数据，既是索引也是物理存储)，后者的主键索引只存储地址。
> 
> 我也是网上查的，欢迎指正 :)

明白您的意思了。
我误以为MyISAM叶子结点的地址是连续存储的，所以叶子结点所指向的数据的地址也是连续存储的，其实不是的。所以不管是主键索引还是二级索引都是随机IO，我赞同您的观点。
谢谢您的耐心解答！也希望大佬继续输出高质量的博文
